src/sdam/monitor.rs

```diff
@@ -21,6 +27,16 @@ pub(super) const DEFAULT_HEARTBEAT_FREQUENCY: Duration = Duration::from_secs(10)
 
 pub(crate) const MIN_HEARTBEAT_FREQUENCY: Duration = Duration::from_millis(500);
 
+macro_rules! emit_event {
```

Comment:
I needed to use a macro here instead of a method because the borrow checker does not allow you to pass an immutable reference to something as an argument while you hold an immutable reference to it. In this case, we're holding a mutable reference to `self` on in `perform_is_master` when matching on `self.connection` in the `Some` branch. Because macros are just expanded at the call site, we can bypass this rule by using a macro rather than a method on `HeartbeatMonitor` (or a method that takes references to `HeartbeatMonitor`'s fields) which would require an immutable reference.